graspologic/cluster/divisive_cluster.py

```diff
@@ -0,0 +1,272 @@
+        Keyword arguments (except min_components and max_components) for chosen
+        clustering method.
+
+    Attributes
+    ----------
+    model_ : GaussianMixture or KMeans object
+        Fitted clustering object based on which `cluster_method` was used.
+
+    See Also
+    --------
+    graspologic.cluster.AutoGMMCluster
+    graspologic.cluster.KMeansCluster
+
+    Notes
+    -----
+    This algorithm was strongly inspired by maggotcluster, a divisive
+    clustering algorithm in https://github.com/neurodata/maggot_models
+
+    References
+    ----------
+    .. [1]  Athey, T. L., & Vogelstein, J. T. (2019).
+            AutoGMM: Automatic Gaussian Mixture Modeling in Python.
+            arXiv preprint arXiv:1909.02688.
+    """
+
+    def __init__(
+        self,
+        cluster_method="gmm",
+        min_components=1,
+        max_components=2,
+        cluster_kws={},
+        min_split=1,
+        max_level=4,
+        delta_criter=0,
+    ):
+
+        _check_common_inputs(min_components, max_components, cluster_kws)
+
+        if cluster_method not in ["gmm", "kmeans"]:
+            msg = "clustering method must be one of"
+            msg += "{gmm, kmeans}"
+            raise ValueError(msg)
+
+        if delta_criter < 0:
+            raise ValueError("delta_criter must be non-negative")
+
+        self.parent = None
+        self.min_components = min_components
+        self.max_components = max_components
+        self.cluster_method = cluster_method
+        self.cluster_kws = cluster_kws
+        self.min_split = min_split
+        self.max_level = max_level
+        self.delta_criter = delta_criter
+
+    def fit(self, X):
+        """
+        Fits clustering models to the data as well as resulting clusters
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+        self.fit_predict(X)
+        return self
+
+    def fit_predict(self, X):
+        """
+        Fits clustering models to the data as well as resulting clusters
+        and using fitted models to predict a hierarchy of labels
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+
+        Returns
+        -------
+        labels : array_label, shape (n_samples, n_levels)
+        """
+        X = check_array(X, dtype=[np.float64, np.float32], ensure_min_samples=1)
+
+        if self.max_components > X.shape[0]:
+            msg = "max_components must be >= n_samples, but max_components = "
+            msg += "{}, n_samples = {}".format(self.max_components, X.shape[0])
+            raise ValueError(msg)
+
+        labels = self._fit(X)
+        # delete the last column if predictions at the last level
+        # are all zero vectors
+        if (labels.shape[1] > 1) and (np.max(labels[:, -1]) == 0):
+            labels = labels[:, :-1]
```

Comment:
@tliu68 can you explain? when will they all be 0 vector?

Reply:
part of why i bring this up - I am wondering whether we have some confusing behavior with regard to the leaf nodes. i.e. some leaf nodes could have no `.model_` and others could have a 1-cluster `.model_`. Is that what causes the thing you have to account for above?

Reply:
1. at line 235 in `_fit` (`child_labels = dc._fit(new_X)`), if the prediction on node `dc` is a 0 vector, `child_labels` here is a 0 vector. If all nodes at the lowest level, i.e., all nodes with `depth = root.height`, have predictions of a 0 vector, the last column of the labels will be all 0s. 
Alternatively, we could add an if condition like `if np.max(child_labels) != 0` before the while loop next line to prevent adding this 0 vector to current `labels`. 
I honestly don't know which one is better code-wise or probably there is another better way. Both seem to be minor adjustment. I kind of think of this step as "cleaning up" the label output, so went w/ the former one to get rid of the redundant column in the end. I'm happy to do it differently if it seems weird to you.

1. Yes, not all nodes have `.model_` currently. The ones that did not pass the if conditions on lines 230-234 won't have it. We could add a sudo pred of 0 vector before the if conditions like I did at line 198
(not sure if those line #s are hard to find but I couldn't find a way to reference them, sorry about that...)